### PR TITLE
IGNITE-20323 : SQL hint for join type.

### PR DESCRIPTION
**Pull request by Vladsz83**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+                for (HintDefinition prevTblHint : prevTblHints) {
+                    boolean prevHintIsDisable = !HINTS.containsKey(prevTblHint);
+
+                    if (prevHintIsDisable)
+                        ++disableCnt;
+
+                    boolean opposite = prevTblHint != curHintDef && (!prevHintIsDisable || !curHintIsDisable)
+                        && (prevTblHint == HINTS.get(curHintDef) || curHintDef == HINTS.get(prevTblHint)
+                        || (!curHintIsDisable && HINTS.get(prevTblHint) != null));
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java:127]: Looks overcomplicated. At least `HINTS.get(prevTblHint) != null` can be simplified to `!prevHintIsDisable', perhaps something else.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    /** */
+    protected static Set<String> joinTblNames(Join join) {
+        Set<String> res = new LinkedHashSet<>();
+
+        for (RelNode in : join.getInputs()) {
+            if (in instanceof RelSubset) {
+                for (RelNode sub : ((RelSubset)in).getRels()) {
+                    if (sub instanceof TableScan)
+                        res.add(last(((TableScan)sub).getTable().getQualifiedName()));
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java]: Aliases aren't available at the rules phase. There is no examples in Calcite too. But there is table hints. My bad, I missed them. I prepared https://github.com/apache/ignite/pull/11130 . It holds an example how table hints might work `JoinTypeHintPlannerTest.testTableHints()`:
`SELECT A2.A, T3.V3, T1.V2 FROM (SELECT 1 AS A, 2 AS B) A2 JOIN TBL3 /*+ MERGE_JOIN */ T3 ON A2.B=A2.B JOIN TBL1 /*+ NL_JOIN */ T1 on T3.V3=T1.V1 where T1.V2=5`
I propose to file a dedicated ticket for the table hints. We might use them for INDEX, ORDERED and join type hints.

[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java]: Here is the ticket for table hints: https://issues.apache.org/jira/browse/IGNITE-21162

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+                continue;
+            }
+
+            for (String tbl : matchedTbls)
+                hintedTables.computeIfAbsent(tbl, t -> new ArrayList<>()).add(curHintDef);
+
+            // This join type is directyly disabled or other join type is forced.
+            if (curHintIsDisable && curHintDef == knownDisableHint || !curHintIsDisable && knownForceHint != curHintDef)
+                ruleDisabled = true;
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java:144]: We didn't processed all the hints yet. We should log or next unapplied hints to HintUtils.skippedHint()

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        return ruleDisabled;
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java:147]: Same. if might be true if we logs all unapplied hints to HintUtils.skippedHint()

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**